providers/namecheap/namecheap.go

```diff
@@ -18,9 +20,136 @@ type Namecheap struct {
 
 func init() {
 	providers.RegisterRegistrarType("NAMECHEAP", newReg)
+	providers.RegisterDomainServiceProviderType("NAMECHEAP", newDsp)
 }
 
-func newReg(m map[string]string) (providers.Registrar, error) {
+func newDsp(conf map[string]string, metadata json.RawMessage) (providers.DNSServiceProvider, error) {
+	return newProvider(conf, metadata)
+}
+
+func newReg(conf map[string]string) (providers.Registrar, error) {
+	return newProvider(conf, nil)
+}
+
+func splitDomain(domain string) (string, string) {
+	ds := strings.Split(domain, ".")
+	dl := len(ds)
+	tld := ds[dl-1]
+	sld := ds[dl-2]
+	return sld, tld
+}
+
+func (n *Namecheap) GetDomainCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+
+	sld, tld := splitDomain(dc.Name)
+	records, err := n.client.DomainsDNSGetHosts(sld, tld)
+	if err != nil {
+		return nil, err
+	}
+
+	var actual []*models.RecordConfig
+	for _, r := range records.Hosts {
+		if r.Type == "SOA" || r.Type == "NS" {
+			continue
+		}
+
+		if r.Name == "@" {
+			r.Name = dc.Name
+		} else {
+			r.Name = r.Name + "." + dc.Name
+		}
+
+		rec := &models.RecordConfig{
+			NameFQDN: r.Name,
+			Type:     r.Type,
+			Target:   r.Address,
+			TTL:      uint32(r.TTL),
+			Priority: uint16(r.MXPref),
+			Original: r,
+		}
+		actual = append(actual, rec)
+
+	}
+
+	differ := diff.New(dc)
+	_, create, delete, modify := differ.IncrementalDiff(actual)
+
+	// because namecheap doesn't have selective create, delete, modify,
+	// we bundle them all up to send at once.  We *do* want to see the
+	// changes though
+
+	changes := false
+	desc := ""
```

Comment:
Yeah, I can see that - I completely stole this section from the gandi provider, so the same can apply there ;)